helix-core/src/main/java/org/apache/helix/controller/stages/MessageOutput.java

```diff
@@ -64,6 +64,18 @@ public void addMessages(String resourceName, Partition partition,
     return Collections.emptyList();
   }
 
+  public Map<Partition, List<Message>> getResourceMessages(String resourceName) {
+    Map<Partition, List<Message>> map = _messagesMap.get(resourceName);
+    if (map != null) {
+      return map;
+    }
+    return Collections.emptyMap();
+  }
+
+  public void addResourceMessages(String resourceName, Map<Partition, List<Message>> resourceMessages) {
```

Comment:
The usage is like this.
```
      MessageOutput output = new MessageOutput();
...
      try {
        Map<Partition, List<Message>> resourceMessages =
            computePerReplicaPartitionState(idealState, currentStateOutput,
                selectedMessage.getResourceMessages(resourceName), resourceMap.get(resourceName),
                bestPossibleStateOutput, dataCache,
                throttleController, retracedPartitionsState, throttledRecoveryMsg, throttledLoadMsg);
        output.addResourceMessages(resourceName, resourceMessages);
        retracedResourceStateMap.setState(resourceName, retracedPartitionsState);
```

if we remove this one, we have to expand this line `output.addResourceMessages(resourceName, resourceMessages);` to a loop, which would be less efficient. 

This is pipeline core logic, saving a few cpu cycle seems to me is worthy. Let me know what is your take?

Reply:
I understand your concern. One easy way to avoid this performance concern, while avoiding breaking the clean API design, is pass the output into the computePerReplicaPartitionState() methods. And inside the method, update the messages gradually use the existing API.

Reply:
changed.